gh-51067: Add `remove()` and `repack()` to `ZipFile`

danny0838 opened the pull request:
This is a revised version of [PR #103033](https://github.com/python/cpython/pull/103033), implementing two new methods in `zipfile.ZipFile`: `remove()` and `repack()`, as suggested in [this comment](https://github.com/python/cpython/pull/103033#issuecomment-1490705652).

## Features

### `ZipFile.remove(zinfo_or_arcname)`

* Removes a file entry (by providing a `str` path or `ZipInfo`) from the central directory.
* If there are multiple file entries with the same path, only one is removed when a `str` path is provided.
* Returns the removed `ZipInfo` instance.
* Supported in modes: `'a'`, `'w'`, `'x'`.

### `ZipFile.repack(removed=None)`

* Physically removes stale local file entry data that is no longer referenced by the central directory.
* Shrinks the archive file size.
* If `removed` is passed (as a sequence of removed `ZipInfo`s), only their corresponding local file entry data are removed.
* Only supported in mode `'a'`.

## Rationales

### Heuristics Used in `repack()`

Since `repack()` does not immediately clean up removed entries at the time a `remove()` is called, the header information of removed file entries may be missing, and thus it can be technically difficult to determine whether certain stale bytes are really previously removed files and safe to remove.

While local file entries begin with the magic signature `PK\x03\x04`, this alone is not a reliable indicator. For instance, a self-extracting ZIP file may contain executable code before the actual archive, which could coincidentally include such a signature, especially if it embeds ZIP-based content.

To safely reclaim space, `repack()` assumes that in a normal ZIP file, local file entries are **stored consecutively**:
- File entries must not overlap.
  - If any entry’s data overlaps with the next, a `BadZipFile` error is raised and no changes are made.
- There should be no extra bytes between entries (or between the last entry and the central directory):
  1. Data before the first referenced entry is removed only when it appears to be a sequence of consecutive entries with no extra following bytes; extra preceeding bytes are preserved.
  1. Data between referenced entries is removed only when it appears to be a sequence of consecutive entries with no extra preceding bytes; extra following bytes are preserved.

Check the doc in the source code of `_ZipRepacker.repack()` (which is internally called by `ZipFile.repack()`) for more details.

### Supported Modes

There has been opinions that a repacking should support mode `'w'` and `'x'` (e. g. https://github.com/python/cpython/issues/51067#issuecomment-1093479250).

This is **NOT** introduced since such modes do not truncate the file at the end of writing, and won't really shrink the file size after a removal has been made. Although we do can change the behavior for the existing API, some further care has to be made because mode `'w'` and `'x'` may be used on an unseekable file and will be broken by such change. OTOH, mode `'a'` is not expected to work with an unseekable file since an initial seek is made immediately when it is opened.

----

<!-- gh-issue-number: gh-51067 -->
* Issue: gh-51067
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--134627.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

danny0838 committed:
Add `remove()` and `repack()` to `ZipFile`

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

danny0838 committed:
Fix and optimize test code

danny0838 committed:
Handle common setups with `setUpClass`

danny0838 committed:
Add tests for mode `w` and `x` for `remove()`

sharktide reviewed (commented):
No one wants to read a PR description that is a mile wrong :)

sharktide commented:
Better

sharktide reviewed (changes requested):
It probably would be better to raise an attributeError instead of a valueError here since you are trying to access an attribute a closed tarfile doesn’t have

danny0838 commented:
> It probably would be better to raise an attributeError instead of a valueError here since you are trying to access an attribute a closed zipfile doesn’t have

This behavior simply resembles `open()` and `write()`, which raises a `ValueError` in various cases. Furthermore there has been a change from raising `RuntimeError` since Python 3.6:

> Changed in version 3.6: Calling [open()](https://docs.python.org/3.13/library/zipfile.html#zipfile.ZipFile.open) on a closed ZipFile will raise a [ValueError](https://docs.python.org/3.13/library/exceptions.html#ValueError). Previously, a [RuntimeError](https://docs.python.org/3.13/library/exceptions.html#RuntimeError) was raised.

> Changed in version 3.6: Calling [write()](https://docs.python.org/3.13/library/zipfile.html#zipfile.ZipFile.write) on a ZipFile created with mode 'r' or a closed ZipFile will raise a [ValueError](https://docs.python.org/3.13/library/exceptions.html#ValueError). Previously, a [RuntimeError](https://docs.python.org/3.13/library/exceptions.html#RuntimeError) was raised.

danny0838 commented:
Nicely inform @barneygale, @merwok, @wimglenn about this PR. This should be more desirable and flexible than the previous PR, although cares must be taken as there might be a potential risk on the algorithm about reclaiming spaces.

The previous PR is kept open in case some folks are interested in it. Will close it either one is accepted.

danny0838 committed:
Introduce `_calc_initial_entry_offset` and refactor

danny0838 committed:
Optimize `_calc_initial_entry_offset` by introducing cache

danny0838 committed:
Introduce `_validate_local_file_entry` and refactor

danny0838 committed:
Introduce `_debug` and refactor

danny0838 committed:
Introduce `_move_entry_data` and rework chunk_size passing

danny0838 committed:
Refactor `_validate_local_file_entry`

danny0838 committed:
Add `strict_descriptor` option

danny0838 committed:
Fix and improve validation tests

- Separate individual validation tests.
- Check underlying repacker not called in validation.
- Use `unlink` to prevent FileNotFoundError.
- Fix mode 'x' test.

danny0838 committed:
Use `zinfo` rather than `info`

danny0838 committed:
Raise on overlapping file blocks

danny0838 committed:
Remove obsolete NameToInfo updating

danny0838 committed:
Rework writing protection

- Set `_writing` to prevent `open('w').write()` during repacking.
- Move the protection logic to `ZipFile.repack()`.

danny0838 committed:
Update doc

sharktide reviewed (changes requested):
FYI: it would speed up the core review and merge if you could please shorten your PR description.

sorry to be annoying :)

merwok commented:
Why do you say that? The description gives context to reviewers. (The problem or request should be described in the issue, but the PR description is appropriate to give details on the implementation of the solution.)

When merging, the core dev will put an appropriate commit message.

emmatyping commented:
@sharktide I understand you are trying to be helpful, but please do not instruct contributors on process, especially if that process is not written down somewhere. There's no requirement for a short PR description, in fact, the long description here is a benefit in my opinion (I would like to review but there is a lot of context to consider!).

sharktide commented:
Sorry! I didn’t type that. I stepped away from the laptop and my friend who has a bad sense of humor but knows his way around GitHub typed that. Deleting those comments. I just found out from these emails!

danny0838 commented on Doc/library/zipfile.rst:
```diff
@@ -518,6 +518,47 @@ ZipFile Objects
+   referenced, shrinking the ZIP file size.
+
+   ``strict_descriptor=True`` can be provided to skip the slower scan for an
+   unsigned data descriptor (deprecated in the latest ZIP specification and is
```
This is to inform the reader whether a negative impact is likely to happen, under the presumption that most people don't know much about unsigned data descriptors for ZIP format.

I would keep this unless a better solution (such as an adequate rephrasing) has come out.

danny0838 committed:
Support `repack(removed)`

sharktide reviewed (approved):
> Perhaps ZipFile could track the offsets of removed entries and use that information to allow for repack to handle arbitrary zips?

Maybe, but I think that would just add extra complexity and a whole new plethora of issues. LGTM so far!

@danny0838 Thanks for the PR!

danny0838 committed:
Fix typo

danny0838 committed:
Add test for bytes between file entries

danny0838 committed:
Check `testzip()` after zip file closed

danny0838 committed:
Fix bytes between entries be removed when `removed` is passed

danny0838 commented:
> @danny0838 thank you for sticking with this issue! Would love to see removal support for zipfiles.
> 
> I'm a bit concerned about `repack` being overly restrictive. Perhaps ZipFile could track the offsets of removed entries and use that information to allow for repack to handle arbitrary zips? In this scenario, repack would scan for local file header signatures, and check the offset against the central directory and list of deleted ZipInfos. If it is in neither, data is left as is and scanning continues. If it is in the deleted list, the file is dropped, if it is in the central directory, then it is kept.

Added support of `remove(removed)` to pass a sequence of removed `ZipInfo`s, and reclaims space only for corresponding local file headers. So one can do something like:

    with zipfile.ZipFile('archive.zip', 'a') as zh:
        zinfos = [zh.remove(n) for n in zh.namelist() if n.startswith('folder/')]
        zh.repack(zinfos)

Though this should do the same thing in most cases, except for being more performant by eliminating the need to scan file entries in the bytes.

danny0838 committed:
Fix bad test code

danny0838 committed:
Revise docstring

danny0838 committed:
Add `tearDown` for tests

danny0838 committed:
Rename methods and parameters

danny0838 committed:
Adjust parameter order

danny0838 committed:
Optimize code and revise comment

- According to ZIP spec, both uncompressed and compressed size should be 0xffffffff when zip64 is used.

danny0838 committed:
Rework `_validate_local_file_entry_sequence` to return size or None

danny0838 committed:
Improve debug for `_ZipRepacker.repack()`

danny0838 committed:
Rework `_validate_local_file_entry_sequence` to allow passing no `checked_offsets`

danny0838 committed:
Strip only entries immediately following a referenced entry

- The previous implementation might cause [archive decryption header] and/or [archive extra data record] preceeding [central directory] be stripped.

danny0838 committed:
Introduce `_scan_data_descriptor_no_sig_by_decompression`

danny0838 committed:
Adjust method names

danny0838 committed:
Add memory usage test

danny0838 committed:
Fix rst

danny0838 commented:
Is there still any problem about this PR?

Here are something that may need further consideration/discussion:


## 1. Should `strict_descriptor` option for `repack()` default to True or False?

### Summary of trade-offs

| Option | Pros ✅ | Cons ❌ |
| ------ | ------- | ------- |
| `strict_descriptor=True` | - Correctly strips any entry with an unsigned data descriptor<br>- Better strict to ZIP spec | - \~150× slower in worst cases<br>- Might open a hole for DoS (if attacker crafts offensive entries)<br>- Slightly higher false-positive risk on random bytes |
| `strict_descriptor=False` | - Much faster<br>- Safer against DoS scenarios<br>- Lower false-positive risk | - Cannot strip unsigned descriptors<br>- Less strict to ZIP spec (but doesn't violate it) |

### Background

When a local file entry has the flag bit indicating usage of data descriptor:
1. This method first attempts to scan for a signed data descriptor.
2. If no valid one is found:
   1. For supported compression methods (`ZIP_DEFLATED`, `ZIP_BZIP2`, or `ZIP_ZSTANDARD`), it decompresses the data to find its end offset.
   2. Otherwise it performs a byte-by-byte scan for an unsigned data descriptor.

This option only affects case **2.2**, which is used when neither signed descriptor nor decompression-based validation is applicable.

### Performance comparison

Based on the benchmark (see tests in `test_zipfile64.py`):
- 8 GiB `ZIP_STORED` file with signed data descriptor: \~56.7s
- 400 MiB `ZIP_STORED` file with unsigned data descriptor: \~270s

The latter is over **150× slower** due to the byte-by-byte scanning for a valid data descriptor.

This may also raise a security concern since `strict_descriptor=False` may open a path for a DoS Attack (if an attacker crafts a ZIP file with offensive entries).

#### False-positive risk

It's not possible to guarantee the "real" file size of a local entry with an unsigned descriptor without the information from the central directory.

If a local file entry spans 100 MiB, it's theoretically possible that multiple byte ranges (e.g., the first 20 MiB, 30 MiB, etc.) could each appear as valid data + data descriptor segments with differing CRCs and compressed sizes. (Currently, the algorithm validates only the compressed size. Checking for CRCs could reduce false positives but would significantly deteriate performance.)

A false positive should be unlikely to happen in practice. If it were to happen, a stale local file entry is stripped incompletely (e.g. a 30MiB entry be treated as 20MiB, leaving 10MiB random bytes over) and cause following entries not stripped (since the algorithm requires consecutive entries). However, the ZIP file remains uncorrupted.

#### Spec compliance

According to the [ZIP file format specification](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT): Applications SHOULD write signed descriptors and SHOULD support both forms when reading.

Unsigned descriptors are thus considered **legacy**, but it is unclear whether they are still used widely.

`strict_descriptor=True` adheres less strictly to the spec, but does **not** violate it — because stripping is neither reading nor writing, and a suboptimal stripping does not corrupt the ZIP archive.


## 2. Should we also implement `copy()` for `ZipFile`?

Currently, copying an entry within a ZIP file is cumbersome due to the lack of support for simultaneous reading and writing. The implementer must either:

- Read the entire entry and write afterwards (which is memory-intensive and inefficient for large files), or
- Use a temporary file for buffered copying.

Both approaches are more complex and less performant, due to the need to decompress and recompress data.

If would be much more performant and friendly by implementing a `copy()` method, using the similar internal buffered copying technique that `_ZipRepacker` has used.

Additionally, this also opens the door to support an efficient `move()` operation, composed of `copy()`, `remove()`, and optionally `repack()`.

And an additional question: whether this should be included in the current PR, or proposed separately as a follow-up?

gpshead commented:
A higher level question: Why would we want to maintain advanced features that are not used by most people within the standard library at all?  This code existing in the stdlib creates a maintenance burden and is the slowest possible way to get features and their resulting trail of bugfixes to people who might want them. All features have an ongoing cost.

Is there a real _need_ for these zipfile features to not simply be advanced ones available in a PyPI package?  @jaraco FYI

I appreciate the enthusiasm for implementing something interesting. I'm not sure we actually want to maintain that within the CPython project though.  Are there compelling use cases for these features to be part of Python rather than external?

danny0838 commented:
@gpshead Dont?'t you think your question should be raised in the issue thread rather than here? 🫠

gpshead commented:
good point, moved, thanks!  _(where to have what discussions is an area where I consider github's UX to be... not great)_

danny0838 committed:
Add tests for `_ZipRepacker`

danny0838 committed:
Optimize code

danny0838 committed:
Fix `_scan_data_descriptor`

danny0838 committed:
Fix and optimize `_scan_data_descriptor_no_sig`

danny0838 committed:
Fix and optimize `_iter_scan_signature`

danny0838 committed:
Fix `_scan_data_descriptor_no_sig_by_decompression`

danny0838 committed:
Rename `_trace_compressed_block_end`

danny0838 committed:
Remove unneeded import

danny0838 committed:
Add requirements

danny0838 committed:
Fix `_scan_data_descriptor_no_sig_by_decompression` when library not available

danny0838 committed:
Test with pre-calculated CRC

danny0838 committed:
Remove unneeded import

danny0838 committed:
Fix and optimize `repack`

danny0838 commented:
@emmatyping Still pending review... I'd like to know if there is still any problem about this PR?

sharktide reviewed (approved):
I think it’s fine

emmatyping commented:
Unfortunately it's premature to move forward with this PR until the discussion in the issue is resolved. There are open questions about whether we should add this API or not, and if we do what it would look like. That needs to be worked out in the issue before this PR can move forward, sorry!

danny0838 committed:
Remove unneeded catch type

- NotImplementedError is a subclass of RuntimeError.

danny0838 committed:
Patch more explicitly

danny0838 committed:
Remove unneeded variables

danny0838 committed:
Improve dependency check for decompression tests

danny0838 committed:
Refactor and optimize `RepackHelperMixin`

jaraco commented:
> And an additional question: whether this should be included in the current PR, or proposed separately as a follow-up?

This sounds like another feature, but it does sound useful. Perhaps it should have its own issue (if it doesn't already). I'd say let's limit the scope here to addressing the linked issue.

jaraco commented:
> Should `strict_descriptor` option for `repack()` default to True or False?

I don't love that there's two implementations here. My reading from the code is that unsigned descriptors are deprecated, so maybe it would be enough to simply omit support for them, opt for the faster implementation, and (maybe) warn if such a descriptor is encountered that it's unsupported. How prevalent are such descriptors?

danny0838 commented:
> > Should `strict_descriptor` option for `repack()` default to True or False?
> 
> I don't love that there's two implementations here. My reading from the code is that unsigned descriptors are deprecated, so maybe it would be enough to simply omit support for them, opt for the faster implementation, and (maybe) warn if such a descriptor is encountered that it's unsupported. How prevalent are such descriptors?

Actually there are three implementations. 😂

Therefore the brief algorithm/condition for the slow scan is:
1. *removed* is not provided.
1. Has a local file header like byte sequence (with `PK\x03\x04` signature, flag bit 8, and CRC == compress_size == file_size == 0), which should be very unlikely to happen on random bytes.
1. A valid signed data descriptor is not seen.
1. A supported compression method (which can reliably detect stream ending) is not used. (STORED/LZMA) (not sure if LZMA decompressor can be reworked to support it)
1. strict_descriptor==True

It should be quite unlikely to happen even if strict_descriptor==True, but the problem is still that it may be catastrophically slow once it happens, and could be used intentionally and offensively.

The prevalence of the deprecated unsigned data descriptor is hard to tell. Most apps like WinZip, WinRAR, 7z would probably write no data descriptor since they are generally not used in a streaming condition. For streaming cases I think most developers would use the ZIP implementation at their hand, and to answer this question we'd have to check the ZIP implementation of popular programming languages like C, Java, JS, php, C#, etc.

Since this feature has been already implemented, I'd prefer to keep it unless there's a strong reason that it should not exist. Defaulting strict_descriptor to False should be enough if it's not welcome. (A minor reason is that we still cannot clearly say that unsigned data descriptor is not supported even if the slow scan is removed, due to the decompression approach)

danny0838 commented:
Reworked `strict_descriptor=True` to ignore any scan of unsigned data descriptors. (https://github.com/danny0838/zipremove/commit/1a07dffcd50adc772dcfaf1e795d391a63f4e16d)

distantvale commented:
@danny0838 in reply to https://github.com/python/cpython/issues/51067#issuecomment-2985304746 (sorry about fragmenting the discussion) -

I'm generally supportive of the iterative approach of start simple and expand from that.

Having said that, since you already implemented _remove_members, I don't see why not expose it through a public API, given that ZipFile has other examples for functions that receive a list of members (members arg of extractall) - can simply follow the same pattern for consistency.
I do agree it's much more useful - my own local implementation only removes a single member and I do see the performance benefits of having an ability to remove multiple members, as many times I myself need to remove multiple members in many different files and it would be much beneficial.

I'm not fond of the idea of a 2 step removal (remove/repack), since it can leave the file in state that some may see (me included) as corruption - zip files aren't support to have a CD entry without a corresponding LF entry.
I'd rather take an approach that limits certain things (cannot remove a file from a self extracting archive for example),
and I don't see the large benefit of delayed repacking flexibility, seems like rather an edge case (but I might be wrong here).

Also some of the questions (how to handle a folder) are relevant regardless of whether we take the 1 step or 2 step approach.
If the 2 step approach has more support here, I'll take it - we can always improve it in the future, but I'd prefer to keep things simpler when possible.

And last but not least - perfect is the enemy of good - it's not the last python version, and we can always improve as long as the initial version is reasonable, which I believe it is.

danny0838 commented:
@distantvale
> Having said that, since you already implemented _remove_members, I don't see why not expose it through a public API, given that ZipFile has other examples for functions that receive a list of members (members arg of extractall) - can simply follow the same pattern for consistency.

I don't think `extractall` is a good analogy since it's read-only. Instead you should check for `open('w')`, `write`, `write_str`, and `mkdir`, each of which is a simple single entry handler.

There are stil design considerations even for your proposed *members*, for example:
1. Should there be `remove` plus `removeall`, or a `remove` that accepts either a single entry or a list?
2. What to do if identical names are passed? Should they be treated as a set and deduplicated, or be removed for multiple times?
3. What to do if any passed value causes an error (such as non-exist)?

The current implementation of `_remove_members` actually defers such decisions. This is fine for a private helper method but is probably premature for a public API.

> I'm not fond of the idea of a 2 step removal (remove/repack), since it can leave the file in state that some may see (me included) as corruption - zip files aren't support to have a CD entry without a corresponding LF entry.

I don't get this. `remove` makes *a LF entry without a corresponding CD entry*, which is clearly allowed by the ZIP spec and won't be complained by any ZIP app.

> I'd rather take an approach that limits certain things (cannot remove a file from a self extracting archive for example), and I don't see the large benefit of delayed repacking flexibility, seems like rather an edge case (but I might be wrong here).

A delayed repacking isn't necessarily done after the archive is closed, it can also happen simply *after the method is called and returned*.

For example, a repacking after a mixed operations of writing, removing, copying, renaming, and just before archive closing, such as an interactive ZIP file editing tool would do.

It's probably challenging enough to design a one-time `remove` or `removeall` to support all such cases.

> Also some of the questions (how to handle a folder) are relevant regardless of whether we take the 1 step or 2 step approach.

I don't think there's too much need to dig into such details for low level APIs—just let them work simply like other existing methods. It's not the case for a one-time high level `remove`—and maybe `copy` and `rename`—though.

danny0838 commented:
@distantvale
> @danny0838 I can share my own opinions about the desired behavior but of course not everyone would agree:
> 
> > I don't think `extractall` is a good analogy since it's read-only. Instead you should check for `open('w')`, `write`, `write_str`, and `mkdir`, each of which is a simple single entry handler.
> 
> I don't see the need to separate into read/write, as I believe they should all be consistent. If certain functions, such as extract, have 2 versions - for single/multiple files, then it won't be an exception in the API, and if we find it useful, have both.

The truth is that they are not consistent. When you say they should be consistent, do you mean there should be multiple file version for `open('w')`, `write`, `write_str`, and `mkdir`, or there should be no multiple file version for `extractall`, or every of them should be a single method that supports both single and list input?

> > There are stil design considerations even for your proposed _members_, for example:
> > 
> > 1. Should there be `remove` plus `removeall`, or a `remove` that accepts either a single entry or a list?
> 
> I'd follow extract in this case.
> 
> > 2. What to do if identical names are passed? Should they be treated as a set and deduplicated, or be removed for multiple times?
> 
> I might be wrong here, but I think `add` allows adding multiple files with the same name, so in that case I'd remove multiple times. If I'mt wrong about `add`, I'd treat it as a set.
> 

Actually there is no `ZipFile.add`. Do you mean `write` and others?

> > 3. What to do if any passed value causes an error (such as non-exist)?
> 
> Just like in `extractall`, the names must be a subset of `namelist()`.
> 

In `extractall` each input is handled one by one, and any error causes subsequent inputs not handled. However the current `_remove_members` handles removing and repacking at the simultaneously. If any error happens in the middle, the whole repacking is partially done and the archive will be in an inconsistent state.

Likewise, for `extractall` providing duplicated names just extract the same entry to the same filesystem path, which would be a different story for removing.

Anyway, the current `_remove_members` implementation DOESN'T do what you state above. If you favor that approach and really want such behavior, you'd have to work on that PR, doing more tests and feedback (I'm probably not going to keep working on that unless it's the final decision of the issue).
It would also be nice if you can provide the code of the implementation you've been using.

distantvale commented:
> @distantvale
> 
> > @danny0838 I can share my own opinions about the desired behavior but of course not everyone would agree:
> > > I don't think `extractall` is a good analogy since it's read-only. Instead you should check for `open('w')`, `write`, `write_str`, and `mkdir`, each of which is a simple single entry handler.
> > 
> > 
> > I don't see the need to separate into read/write, as I believe they should all be consistent. If certain functions, such as extract, have 2 versions - for single/multiple files, then it won't be an exception in the API, and if we find it useful, have both.
> 
> The truth is that they are not consistent. When you say they should be consistent, do you mean there should be multiple file version for `open('w')`, `write`, `write_str`, and `mkdir`, or there should be no multiple file version for `extractall`, or every of them should be a single method that supports both single and list input?
> 
What I'm saying is that there's already precendence in the package for a function that has 2 versions, so it's not out of the ordinary to add another one.

> > > There are stil design considerations even for your proposed _members_, for example:
> > > 
> > > 1. Should there be `remove` plus `removeall`, or a `remove` that accepts either a single entry or a list?
> > 
> > 
> > I'd follow extract in this case.
> > > 2. What to do if identical names are passed? Should they be treated as a set and deduplicated, or be removed for multiple times?
> > 
> > 
> > I might be wrong here, but I think `add` allows adding multiple files with the same name, so in that case I'd remove multiple times. If I'mt wrong about `add`, I'd treat it as a set.
> 
> Actually there is no `ZipFile.add`. Do you mean `write` and others?
> 
Yup, I mean `write`.

> > > 3. What to do if any passed value causes an error (such as non-exist)?
> > 
> > 
> > Just like in `extractall`, the names must be a subset of `namelist()`.
> 
> In `extractall` each input is handled one by one, and any error causes subsequent inputs not handled. However the current `_remove_members` handles removing and repacking simultaneously. If any error happens in the middle, the whole repacking is left partially done and the archive will be in an inconsistent state, which is unlikely an acceptable consequence.
> 
The behavior of `extractall` might not be aligned with the documentation that states that `members` must be a subset of `namelist()` - hard to say what the intention was, but I think it's a legitimate check.
I agree re. the inconsistent state, and maybe that's the most significant reason to keep the external API simple for now. Once the initial version is release, expanding it to more functionality is easier.

> Likewise, for `extractall` providing duplicated names just extract the same entry to the same filesystem path again; for removing this would be a totally different story.
> 
Yeah I agree, and since multiple files with the same names are not very common, I'd opt for treating `members` as a set.

> Anyway, the current `_remove_members` implementation DOESN'T do what you state above. If you favor that approach and really want such behavior, you'd have to work on that PR, doing more tests and feedback (I'm probably not going to keep working on that unless it's the final decision of the issue). It would also be nice if you can provide the code of the implementation you've been using.

My implementation is based on the original PR, so not much there in regards to these options.

In any case I'd proceed with your approach for the time being.

danny0838 committed:
Update NEWS

danny0838 committed:
Sync with danny0838/zipremove@1691ca25bf971cf1e45d5ed7d22c512636f20cb8

danny0838 committed:
Revise NEWS

danny0838 committed:
Sync with danny0838/zipremove@1843d87b70e6cb129fb55446eaf4486a87d2af4d

danny0838 committed:
Fix timezone related timestamp issue

danny0838 committed:
Simplify tests with data descriptors

danny0838 committed:
Sync with danny0838/zipremove@e79042768f3c2541e0226f6bed3a9ff2ee04fac0

danny0838 committed:
Sync with danny0838/zipremove@87bcdb50411a355d24c35f31dcbe4273c0568cf8

danny0838 committed:
Sync with danny0838/zipremove@6a78bd15de87afde510f8a1b6364365c6e17f252

emmatyping commented on Lib/zipfile/__init__.py:
```diff
@@ -1367,6 +1374,459 @@ def close(self):
+            - Modifies the ZIP file in place.
+            - Updates zfile.start_dir to account for removed data.
+            - Sets zfile._didModify to True.
+            - Updates header_offset and clears _end_offset of referenced
```
It would probably be good to also update `data_offset` if anything before the existing `data_offset` is removed.

danny0838 replied:
> data_offset is the offset to the beginning of the zip content in a file, since there may be other content before the beginning of the first local entry.

As I've pointed out in that comment, data_offset can still be 0 even if there are other content before the beginning of the first local entry, making it a bad indicator.

Without making clear what `data_offset` ought to do and possibly whether it's behavior requires a redesign/change first, we cannot decide how to update it's value here.

danny0838 replied:
Close due to #136950.

emmatyping commented on Lib/zipfile/__init__.py:
```diff
@@ -1367,6 +1374,459 @@ def close(self):
+            entry_size_list.append(entry_size)
+            used_entry_size_list.append(used_entry_size)
+
+        # calculate the starting entry offset (bytes to skip)
+        if removed is None:
+            try:
+                offset = filelist[0].header_offset
+            except IndexError:
+                offset = zfile.start_dir
+            entry_offset = self._calc_initial_entry_offset(fp, offset)
+        else:
+            entry_offset = 0
```
I think you can use zfile.data_offset if it isn't None and fall back to this code?

danny0838 replied:
No, it doesn't work.

`ZipFile.data_offset` is always 0 if the offsets in the central directory are calculated from the starting of the file (i.e. bytes before the first entry is included in the calculation).

Minimal reproducible example:

```
with open('archive.zip', 'wb') as fh:
    with zipfile.ZipFile(fh, 'w') as zh:
        fh.write(b'dummy')
        zh.writestr('file', 'file content')

with zipfile.ZipFile('archive.zip') as zh:
    print(zh.data_offset)
```

This seems to be somehow inconsistent to the documentation of `ZipFile.data_offset`—`The offset to the start of ZIP data from the beginning of the file`, but may depend on how `the start of ZIP data` is interpreted. In some sense it looks like a bug, but I'm not sure as the original issue introducing it (#84481) does not clearly mention what intended use case of the attribute is.

danny0838 replied:
Close due to #136950.

emmatyping commented on Lib/zipfile/__init__.py:
```diff
@@ -1367,6 +1374,459 @@ def close(self):
+                    return entry_size
+        return 0
+
+    def _iter_scan_signature(self, fp, signature, start_offset, end_offset, chunk_size=4096):
```
```suggestion
    def _iter_scan_signature(self, fp, signature, start_offset, end_offset, chunk_size=io.DEFAULT_BUFFER_SIZE):
```
This is a bit more portable and removes a magic number.

emmatyping reviewed (changes requested):
Some thoughts in an initial pass. This is a pretty big change so I haven't had time to go over the repack implementation in detail. Thank you for working on all of these changes. I can see an impressive attention to detail!

danny0838 committed:
Sync with danny0838/zipremove@092f98b4d7b3a0cd335fe4ba64e7090ebb3dc6da

danny0838 commented on Lib/test/test_zipfile/test_core.py:
```diff
@@ -1360,6 +1361,2002 @@ class LzmaWriterTests(AbstractWriterTests, unittest.TestCase):
+        ]
+
+        # suppress duplicated name warning
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
```
Fixed in 0832528

danny0838 commented on Doc/library/zipfile.rst:
```diff
@@ -518,6 +518,66 @@ ZipFile Objects
+   The algorithm assumes that local file entries (and the central directory,
+   which is mostly treated as the "last entry") are stored consecutively:
+
+   #. Data before the first referenced entry is removed only when it appears to
+      be a sequence of consecutive entries with no extra following bytes; extra
+      preceding bytes are preserved.
```
> if the bytes `PK\003\004` appear before the first entry in some other format then a user cannot use repack according to my reading of this

In this case the `PK\003\004` are "extra preceding bytes", and will be preserved. Since they are not "extra following bytes" they won't prevent the repack to happen.

emmatyping replied:
There are many formats that build on top of zip, such as doc, java class files, etc. These have prefixes and you need to be sure what you detect is an unreferenced file entry vs a local file entry magic and noise. As far as I am aware, there's no way to be certain what you have is an actual file entry. So there are potentially cases where the following layout could be a misinterpretation of a prefix to a zip file, and repack would incorrectly remove data from that prefix.

```
[unreferenced local file entry]
[PK\003\004 and noise]
[first local file entry]
```

gpshead replied:
Given that heuristics and data scanning are involved I think it is worth while to add a `.. note::` to the repack docs here stating that it (1) cannot be guaranteed" safe to use on untrusted zip file inputs or (2) does not guarantee that zip-like archives such as those with executable data prepended will survive unharmed.

Realistically I think people with (2) style formats should know this but it is good to set expectations.

I'm asking for (1) preemptively because we're basically guaranteed to get security "vulnerability" reports about all possible API behaviors today _[I'm on the security team - we really do]_ so pre-documenting the thing makes replying to those easier. 😺
Given we are never going to be able to prevent all such DoS and zipbomb, frankenzip, or multiple-interpretations-by-different-tooling-zip format reports given the zip format definition combined with the collective behavior of the worlds implementations is so... fuzzy.

> can you provide a good example or test case that normal bytes be mis-interepeted and falsely removed as local file entries before the first referenced local file entry?

That's exactly the kind of thing someone would do in a future security@ report. :P  We can stay ahead of that by at least not offering a guarantee of safety for this API.  The first innocent real world starting points that come to mind are zip files embedded stored within zip files.  And file formats involving multiple zip files within them where only one of them appears at the end of the file and thus acts like a zip to _normal_ zip file tooling such as `zipfile` seeking out the end of file central directory.  Those might not trigger a match in your logic as is, but work backwards from the logic and there may be ways to construct some that could?  I'm just suggesting we don't try to overthink our robustness and make guarantees here.

danny0838 replied:
For (1): not only `ZipFile.repack` but the whole `zipfile` package is not guaranteed to be safe on an untrusted ZIP file. If a note or warning is required, it probably should be for the whole package rather than for this method only.

For (2): this is what `repack` want to at least guarantee. The current algorithm should be quite safe against a false positive as it's very unlikely that random binary bytes would happen to form a local file header magic and happen to have its "entry length" ends exactly at the position of the first referenced local file header.

This can even be improved by providing an option to check for CRC when validating every seemingly like "local file entry". Though it would impact performance significantly and I don't think it worth.

A prepended zip should be safe since it has a central directory, which will be identified as "extra following bytes" and skipped from stripping. Unless the zip is abnormal, e.g. having the last local file entry overlapping in the central comment and thus having no additional bytes after its end.

A zip embedded as the content of a member is also 100% safe since the algorithm won't strip anything inside a local file entry.

A zip embedded immediately after a local file entry will be falsely stripped, but it's explicitly precluded by the documented presumption that "local file entries are stored consecutively", and should be something unlikely to happen on a normal zip-like file.

Given that the current documentation already explains its assumption and algorithm, I expect that the developer be able to estimate the risk on his own. Although it's not 100% safe, worrying about this may be something like worrying about a repository breaking due to SHA-1 collision when using Git. I agree that it would be good to set a fair expectation on the heuristics based algorithm, but I also I don't want to give an impression that the algorithm is something fragile and could easily blow on a random input. Don't you think it's overkill for Git to show a big warning saying that it doesn't guarantee your data won't break accidentally?

Anyway, I'm open to this. It's welcome if someone can provide a good rephasing or note without such issues.

danny0838 commented on Doc/library/zipfile.rst:
```diff
@@ -518,6 +518,66 @@ ZipFile Objects
+   Removes a member from the archive.  *zinfo_or_arcname* may be the full path
+   of the member or a :class:`ZipInfo` instance.
+
+   If multiple members share the same full path, only one is removed when
+   a path is provided.
```
Actually the current implementation is definite that the one mapped by `ZipFile.getinfo(name)` will get removed and the last one in the filelist with the same name (if exists) will be the new mapped one.

This is similar to what will be mapped by `ZipFile.getinfo(name)` if there are multiple zinfos with same name. The current implementation is always the last one in the filelist, though it's also undocumented.

The question is that should we document the definite behavior or state that it's actually undefined and the current behavior should not be relied on? Before the question is solved I would just keep the current statement.

gpshead replied:
I haven't gone back and looked, but I have a vague recollection that multiple zip entries with the same name is/was a "normal" zip file legacy-ish "feature" as that was how replacing the contents of one zip file member was implemented in cases where the entire thing cannot be rewritten as it could be done solely by rewriting the end of the file and central directory.

danny0838 replied:
@gpshead This is the convention of many ZIP tools, including Python, at least currently. Unfortunately it's not clearly documented in the ZIP spec.

danny0838 commented on Doc/library/zipfile.rst:
```diff
@@ -518,6 +518,66 @@ ZipFile Objects
+   If multiple members share the same full path, only one is removed when
+   a path is provided.
+
+   This does not physically remove the local file entry from the archive.
+   Call :meth:`repack` afterwards to reclaim space.
```
Local file entry (of the member) clearly means the content and metadata in a ZIP file. I agree that "local file entry" may be somehow too technical for casual developers to understand, but "the contents of the local file entry" is even more unnatural to me. I would keep the current statement unless a better rephrasing comes out.

merwok replied:
Something like: This removes the local file entry in the ZIP metadata, but does not remove the file contents from the archive.

danny0838 replied:
Maybe move the details to a note and provide more description besides "central directory" and "local file entry"?

```rst
.. method:: ZipFile.remove(zinfo_or_arcname)

   Removes a member from the archive.  *zinfo_or_arcname* may be the full path
   of the member or a :class:`ZipInfo` instance.  If multiple members share the
   same full path and the path is provided, only one of them is removed.

   This does not physically remove the local file entry from the archive.
   Call :meth:`repack` afterwards to reclaim space.

   The archive must be opened with mode ``'w'``, ``'x'`` or ``'a'``.

   Returns the removed :class:`ZipInfo` instance.

   Calling :meth:`remove` on a closed ZipFile will raise a :exc:`ValueError`.

   .. note::
      This method removes only the member's entry from the central directory,
      which is a safe and quick operation and makes it inaccessible to most
      tools.  Its local file entry is preserved, meaning the content and
      metadata are still present and recoverable using certain file recovery
      techniques.  Call :meth:`repack` afterwards to remove the local file
      entry and reclaim space.

   .. versionadded:: next

```

merwok commented on Doc/library/zipfile.rst:
```diff
@@ -518,6 +518,66 @@ ZipFile Objects
+   *chunk_size* may be specified to control the buffer size when moving
+   entry data (default is 1 MiB).
+
+   The archive must be opened with mode ``'a'``.
```
Yes it’s fine. Strict consistency is not a goal that trumps others 🙂

danny0838 committed:
Revise doc for `remove`

danny0838 committed:
Revise doc for `repack`

danny0838 committed:
Update `data_offset`

danny0838 commented on Doc/library/zipfile.rst:
```diff
@@ -518,6 +518,67 @@ ZipFile Objects
+   #. Entries must not overlap. If any entry's data overlaps with another, a
+      :exc:`BadZipFile` error is raised and no changes are made.
+
+   When scanning, setting ``strict_descriptor=True`` disables detection of any
```
This is exactly one of the open question(<https://github.com/python/cpython/pull/134627#issuecomment-2923807248>, <https://github.com/python/cpython/pull/134627#issuecomment-2974467284>).

The current quick decision is primarily since it adheres better to the spec and most Python stdlib tend to prioritize compatibility than performance. E.g. `json.dump` with `ensure_ascii=True` and `http.server` with HTTP version 1.0. But it's not solid and can be changed, based on a vote or something?

danny0838 commented on Doc/library/zipfile.rst:
```diff
@@ -518,6 +518,66 @@ ZipFile Objects
+   If *removed* is not provided, the archive is scanned to identify and remove
+   local file entries that are no longer referenced in the central directory.
+   The algorithm assumes that local file entries (and the central directory,
+   which is mostly treated as the "last entry") are stored consecutively:
```
What about simply **non-referenced** local file entries?

distantvale replied:
ambiguous imo

danny0838 committed:
Revise doc for `repack`

danny0838 committed:
Revise doc for `repack`

danny0838 committed:
Sync with danny0838/zipremove@8bedf7c9b891acadc3393d2f1267b78bd9b5a49a

danny0838 committed:
Sync with danny0838/zipremove@86a240bf019fe9212b1e72c963306186163fb8b8

danny0838 committed:
Sync with danny0838/zipremove@8ad341bdd28a78033d7111a0532cb4714349aee3

danny0838 commented:
@emmatyping The review is still awaiting for nearly a year. Is there still any major issue preventing this PR to be merged in Python 3.15?

gpshead committed:
Merge remote-tracking branch 'origin/main' into gh-51067-2

gpshead committed:
docs: clarify unspecified multiple entry removal, suggest ZipInfo

gpshead committed:
whatsnew entry

gpshead committed:
gh-51067: Default ZipFile.repack() to strict_descriptor=True

Locating an unsigned data descriptor for a STORED or encrypted entry
requires a byte-by-byte scan that can be ~100-1000x slower and is a
potential denial-of-service vector on untrusted input. Default
strict_descriptor to True so repack() detects only the signed data
descriptors written by Python and other modern tools, leaving the slow
scan as an opt-in.

Also explain data descriptors and the signed/unsigned distinction in
the docs, and add coverage for the new default (unit, encrypted, and an
end-to-end repack test across all compression methods).

gpshead committed:
gh-51067: Document reliable space reclamation in ZipFile.repack()

Clarify that passing removed= reclaims a member's space regardless of
how its local file entry was written, while the scan used when removed
is omitted may leave entries in place (e.g. unsigned data descriptors
under the default strict_descriptor=True). Adjust the remove() note so
it no longer implies a bare repack() always removes the data, and add a
short remove-then-repack(removed) example.

gpshead committed:
gh-51067: Bound _validate_local_file_entry to the scanned gap

Without this check a stale local header whose compress_size points past
the next referenced entry would make _validate_local_file_entry_sequence
report more strippable bytes than the gap holds.  The move loop would
then over-advance entry_offset, drive a later header_offset negative,
and fail in fp.seek() with the archive partially rewritten.

gpshead committed:
gh-51067: Guard _copy_bytes against short reads

Avoids a tight loop if a future caller (or a truncated archive) ever
asks to copy past EOF.

gpshead committed:
gh-51067: Spell out ZipFile.repack() keyword arguments

Replace **opts passthrough with the documented keyword-only signature
so help()/inspect show the real parameters and the internal debug=
hook is not reachable from public API.

gpshead committed:
gh-51067: Move NEWS entry to Library section

zipfile is a stdlib module, not core/builtins.

gpshead committed:
gh-51067: Fix docstring typos in _ZipRepacker

gpshead committed:
gh-51067: Drop unreachable struct.error handler in no-sig DD scan

The loop bound guarantees each slice is exactly dd_size bytes.

gpshead committed:
gh-51067: Use io.DEFAULT_BUFFER_SIZE in no-sig DD scan

Consistent with the other _ZipRepacker scanners.

gpshead committed:
gh-51067: Replace ComparableZipInfo class with a function

The previous spelling overrode __new__ to return a dict, which works but
reads as a class that isn't one.  Rename to comparable_zinfo() per PEP 8.

gpshead committed:
gh-51067: Add _REPACK_CHUNK_SIZE constant

Single source of truth for the 1 MiB default used by both
ZipFile.repack() and _ZipRepacker.

gpshead committed:
gh-51067: Add tests for entry-size overshoot and _copy_bytes EOF

Covers the new bounds check in _validate_local_file_entry and the
short-read guard in _copy_bytes.

gpshead reviewed (approved):
I applied a variety of minor edge case fixes for robustness and some code style cleanups.
